Source/JavaScriptCore/builtins/ModuleLoader.js

```diff
@@ -23,586 +23,642 @@
-        var dependenciesMap = moduleRecord.dependenciesMap;
-        var requestedModules = this.requestedModules(moduleRecord);
-        var dependencies = @newArrayWithSize(requestedModules.length);
-        for (var i = 0, length = requestedModules.length; i < length; ++i) {
-            var depName = requestedModules[i];
-            var depKey = this.resolve(depName, key, fetcher);
-            var depEntry = this.ensureRegistered(depKey);
-            @putByValDirect(dependencies, i, depEntry);
-            dependenciesMap.@set(depName, depEntry);
+    @assert(state.isLoading);
+
+    if (entry.status === @ModuleStatusNew && !state.visited.@has(entry)) {
+        state.visited.@add(entry);
+        state.pendingModulesCount += entry.requestedModules.length;
+
+        var { fetcher } = state;
+        for (var i = 0, { length } = entry.requestedModules; i < length; i++) {
+            const specifier = entry.requestedModules[i];
+            var loadedEntry = entry.loadedModules.@get(specifier);
+            if (loadedEntry) {
+                if (loadedEntry.errorToRethrow) {
+                    state.parseError = loadedEntry.errorToRethrow;
+                }
+
+                this.innerModuleLoading(state, loadedEntry);
+            } else {
+                try {
+                    var key = this.resolve(specifier, entry.key, fetcher);
+                } catch (error) {
+                    // This implements step 9.3 of https://html.spec.whatwg.org/multipage/webappapis.html#creating-a-javascript-module-script
+                    entry.errorToRethrow = error;
+                    throw error;
+                }
+
+                this.hostFetchAndLoadImportedModule(key, entry.requestedModuleParameters[i], fetcher, state).then(loadedEntry => {
+                    // This implements step 1 of FinishLoadingImportedModule
+                    @assert(!entry.loadedModules.@has(specifier) || entry.loadedModules.@get(specifier) === loadedEntry);
+                    entry.loadedModules.@set(specifier, loadedEntry);
+
+                    // This implements steps 1-2 of ContinueModuleLoading
+                    if (state.isLoading)
+                        this.innerModuleLoading(state, loadedEntry);
+                }, continueModuleLoadingAbrupt);
+            }
         }
-        entry.dependencies = dependencies;
-        entry.module = moduleRecord;
-        @setStateToMax(entry, @ModuleSatisfy);
-        return entry;
-    })();
-    return instantiatePromise;
-}
 
-@linkTimeConstant
-function cacheSatisfy(entry)
-{
-    "use strict";
+        if (!state.isLoading)
+            return;
+    }
 
-    @setStateToMax(entry, @ModuleLink);
-    entry.satisfy = (async () => {
-        return entry;
-    })();
+    @assert(state.pendingModulesCount >= 1);
+    state.pendingModulesCount--;
+    if (state.pendingModulesCount === 0) {
+        state.isLoading = false;
+        state.visited.@forEach(loadedEntry => {
```

Comment:
Ditto.

Reply:
`state.visited` is a `Set`, we cannot iterate it otherwise (`for/of` would be *actually* observable).